prometheus/client.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.  // You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+)
+
+var (
+	// defining this global variable will avoid to initialized it each time
+	// and it will crash immediately the server during the initialization in case the version is not well defined
+	requiredVersion = semver.MustParse("2.15.0") // nolint: gochecknoglobals
+)
+
+func buildGenericRoundTripper(connectionTimeout time.Duration) *http.Transport {
+	return &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   connectionTimeout,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		TLSHandshakeTimeout: 30 * time.Second,
+		TLSClientConfig:     &tls.Config{InsecureSkipVerify: true}, // nolint: gas, gosec
+	}
+}
+
+func buildStatusRequest(prometheusURL string) (*http.Request, error) {
+	finalURL, err := url.Parse(prometheusURL)
+	if err != nil {
+		return nil, err
+	}
+
+	// define the path of the buildInfo
+	// using this way will remove any issue that could be caused by a wrong URL set by the user
+	finalURL.Path = "/api/v1/status/buildinfo"
+
+	httpRequest, err := http.NewRequest(http.MethodGet, finalURL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	// set the accept content type
+	httpRequest.Header.Set("Accept", "application/json")
+	return httpRequest, nil
+}
+
+type buildInfoResponse struct {
+	Status    string        `json:"status"`
+	Data      buildInfoData `json:"data,omitempty"`
+	ErrorType string        `json:"errorType,omitempty"`
+	Error     string        `json:"error,omitempty"`
+	Warnings  []string      `json:"warnings,omitempty"`
+}
+
+// buildInfoData contains build information about Prometheus.
+type buildInfoData struct {
+	Version   string `json:"version"`
+	Revision  string `json:"revision"`
+	Branch    string `json:"branch"`
+	BuildUser string `json:"buildUser"`
+	BuildDate string `json:"buildDate"`
+	GoVersion string `json:"goVersion"`
+}
+
+type Client interface {
```

Comment:
The exported interfaces functions and methods should have some doc comments.

Reply:
done